lib/logthrdestdrv.c

```diff
@@ -303,9 +294,9 @@ _perform_flush(LogThreadedDestDriver *self)
    * flush() being called always, even if WORKER_INSERT_RESULT_SUCCESS is
    * returned, in which case batch_size is already zero at this point.
    */
-  if (!self->suspended && self->worker.flush)
+  if (!self->suspended)
     {
-      worker_insert_result_t result = self->worker.flush(self);
+      worker_insert_result_t result = log_threaded_dest_worker_flush(self);
```

Comment:
This change is not an equivalent transformation. In the original version, accepting the batch happens only if flush vfunction is provided. In the new code it happens everytime. This can only be a problem when a driver wants to use the batching framework without specifying a flush. I cannot tell any reasonable usecase for this,  though the API allows to do so.

The cleanest would be to start the function with:
```
if (!self->worker.flush)
  return;
```

Reply:
@furiel: I think it is equivalent. Just check the implementation of `log_threaded_dest_worker_flush`.
(and IMHO it is better as checking of the existence of a virtual method is moved to the right place)

Reply:
I still think it is not equivalent, at least not mathematically. In the original implementation with self->worker.flush == 0, the whole if block is skipped. In the new implementation, we simulate a dummy flush method by returning `WORKER_INSERT_RESULT_SUCCESS`. But that has a side effect: acknowledging a half-full batch, which the original implementation did not do. Maybe we can reason why it is not a problem (I am not 100% convinced yet), but this is a functional change in the middle of an otherwise innocent-looking refactor patch, which is unfortunate.

Reply:
I agree this is not functionally equivalent, however after giving some thought this should have been the way it was coded originally. I should have added more comments to the commit message, sorry for that.

As for the rationale:

Remember that in earlier versions of this code, on_message_queue_empty() was called if provided and flush superceedes that logic. 

The original on_message_queue_empty() method was returning void, basically assuming that any flush that needed happening indeed happened successfully, there was no way to indicate failure.

With the new flush logic, in this codepath, one of two things could happen:

*1) the flush() method is provided*

In this case  we work as expected.

*2) flush() is not set*

then I'd expect the driver never return a QUEUED verdict as there's no way to handle queued messages. So
instead of QUEUED it would return a final verdict like SUCCESS or ERROR. In either case the insert() method would already call _process_result(), causing batch_size to be reset to zero. This in turn means that the process_result() here would become a noop, as anything it does is basically dependant on batch_size. (well, it also resets the retries_counter, but again that shouldn't matter that we set that to zero twice).

Reply:
Thanks for the detailed explanation. This was the missing puzzle for me, without flush but returning QUEUED users cannot accept batch. I see `_accept_batch` is not exposed to the `logthrdestdrv.h` header. (Not static function either, maybe worth change it to static later.).